pkg/kubelet/pleg/evented.go

```diff
@@ -0,0 +1,436 @@
+	genericPlegRelistPeriod time.Duration, clock clock.Clock) PodLifecycleEventGenerator {
+	return &EventedPLEG{
+		runtime:                     runtime,
+		runtimeService:              runtimeService,
+		eventChannel:                eventChannel,
+		cache:                       cache,
+		genericPleg:                 genericPleg,
+		eventedPlegMaxStreamRetries: eventedPlegMaxStreamRetries,
+		genericPlegRelistPeriod:     genericPlegRelistPeriod,
+		clock:                       clock,
+	}
+}
+
+// Watch returns a channel from which the subscriber can receive PodLifecycleEvent events.
+// TODO: support multiple subscribers.
+func (e *EventedPLEG) Watch() chan *PodLifecycleEvent {
+	return e.eventChannel
+}
+
+// Relist relists all containers using GenericPLEG
+func (e *EventedPLEG) Relist() {
+	e.genericPleg.Relist()
+}
+
+// Start spawns a goroutine to relist periodically.
+func (e *EventedPLEG) Start() {
+	IsEventedPLEGInUse = true
+	go wait.Forever(e.watchEventsChannel, 0)
+}
+
+// Healthy check if PLEG work properly.
+func (e *EventedPLEG) Healthy() (bool, error) {
+	// GenericPLEG is declared unhealthy when relisting time is more than
+	// the relistThreshold. In case, EventedPLEG is turned on the value of
+	// relistingPeriod, relistingThreshold is adjusted to higher values and
+	// so health of Generic PLEG should take care of the condition checking
+	// higher relistThreshold between last two relist.
+
+	// EventedPLEG is declared unhealthy only if eventChannel is out of capacity.
+	if len(e.eventChannel) >= cap(e.eventChannel) {
+		return false, fmt.Errorf("EventedPLEG: pleg event channel capacity is full with %v events", len(e.eventChannel))
+	}
+
+	timestamp := e.clock.Now()
+	metrics.PLEGLastSeen.Set(float64(timestamp.Unix()))
+	return true, nil
+}
+
+func (e *EventedPLEG) watchEventsChannel() {
+
+	containerEventsResponseCh := make(chan *runtimeapi.ContainerEventResponse, cap(e.eventChannel))
+	defer close(containerEventsResponseCh)
+
+	// Get the container events from the runtime.
+	go func() {
+		numAttempts := 0
+		for {
+			err := e.runtimeService.GetContainerEvents(containerEventsResponseCh)
+			if err == io.EOF {
+				// The runtime service has been closed, exit the goroutine.
+				klog.V(4).Info("Evented PLEG: Runtime service has been closed, exit the goroutine")
+				return
+			}
+
+			if err != nil && numAttempts <= e.eventedPlegMaxStreamRetries {
+				numAttempts++
+				e.Relist() // Force a relist to get the latest container and pods running metric.
+				klog.V(4).Infof("Evented PLEG: Failed to get container events, retrying: %v", err)
+				continue
+			}
+
+			if numAttempts >= e.eventedPlegMaxStreamRetries {
+				klog.ErrorS(err, "Evented PLEG: Failed to get container events")
+				IsEventedPLEGInUse = false
+				// Fall back to Generic PLEG relisting since Evented PLEG is not working.
+				wait.Until(e.Relist, e.genericPlegRelistPeriod, wait.NeverStop)
```

Comment:
If here fallbacks to generic PLEG, is it neccessary to stop `eventedPlegRelistPeriod `Generic PLEG which was started at inital phase? or we would have two relist task with different steps

Reply:
Thanks for pointing it out. This needs to be fixed.

Reply:
@249043822 it's fixed now.